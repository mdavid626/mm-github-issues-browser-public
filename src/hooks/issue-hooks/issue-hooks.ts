import { useQuery, gql } from '@apollo/client';
import { ApolloError } from '@apollo/client/errors';
import { IssueQueryResult, IssuesQueryResult } from '../../types/issue';

const issuesQuery = gql`
  query {
    repository(owner: "facebook", name: "react") {
      issues(last: 20) {
        edges {
          node {
            id
            createdAt
            title
            state
            number
            url
            author {
              login
            }
          }
        }
      }
    }
  }
`;

export const useIssues = (): [
  IssuesQueryResult | undefined,
  boolean,
  ApolloError | undefined
] => {
  const { data, loading, error } = useQuery<IssuesQueryResult>(issuesQuery);
  return [data, loading, error];
};

const issueQuery = gql`
  {
    repository(owner: "facebook", name: "react") {
      issue(number: 26077) {
        id
        createdAt
        title
        state
        number
        stateReason
        url
        bodyText
        author {
          login
        }
      }
    }
  }
`;

export const useIssue = (): [
  IssueQueryResult | undefined,
  boolean,
  ApolloError | undefined
] => {
  const { data, loading, error } = useQuery<IssueQueryResult>(issueQuery);
  return [data, loading, error];
};
