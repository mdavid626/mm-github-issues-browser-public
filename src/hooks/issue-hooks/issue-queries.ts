import { gql } from '@apollo/client';

export const GetIssuesQuery = gql`
  query getIssues($cursor: String, $search: String!) {
    search(query: $search, after: $cursor, type: ISSUE, first: 10) {
      issueCount
      nodes {
        ... on Issue {
          id
          createdAt
          title
          state
          number
          url
          author {
            login
          }
        }
      }
      pageInfo {
        hasNextPage
        endCursor
      }
    }
  }
`;

export const GetIssueQuery = gql`
  query getIssue($issueNumber: Int!) {
    repository(owner: "facebook", name: "react") {
      id
      issue(number: $issueNumber) {
        id
        createdAt
        title
        state
        number
        stateReason
        url
        body
        author {
          login
        }
        comments(first: 3, orderBy: { field: UPDATED_AT, direction: DESC }) {
          nodes {
            id
            createdAt
            body
            author {
              login
            }
          }
          pageInfo {
            hasNextPage
            endCursor
          }
          totalCount
        }
      }
    }
  }
`;

export const GetIssueCommentsQuery = gql`
  query getIssueComments($issueNumber: Int!, $commentsCursor: String) {
    repository(owner: "facebook", name: "react") {
      id
      issue(number: $issueNumber) {
        id
        comments(
          first: 3
          after: $commentsCursor
          orderBy: { field: UPDATED_AT, direction: DESC }
        ) {
          nodes {
            id
            createdAt
            body
            author {
              login
            }
          }
          pageInfo {
            hasNextPage
            endCursor
          }
          totalCount
        }
      }
    }
  }
`;
