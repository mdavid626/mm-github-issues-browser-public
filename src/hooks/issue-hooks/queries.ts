import { gql } from '@apollo/client';

export const GetIssuesQuery = gql`
  query getIssues($cursor: String, $states: [IssueState!]) {
    repository(owner: "facebook", name: "react") {
      id
      issues(
        first: 10
        after: $cursor
        states: $states
        orderBy: { field: CREATED_AT, direction: DESC }
      ) {
        nodes {
          id
          createdAt
          title
          state
          number
          url
          author {
            login
          }
        }
        pageInfo {
          hasNextPage
          startCursor
          endCursor
        }
        totalCount
      }
    }
  }
`;

export const GetIssueQuery = gql`
  query getIssue($issueNumber: Int!) {
    repository(owner: "facebook", name: "react") {
      id
      issue(number: $issueNumber) {
        id
        createdAt
        title
        state
        number
        stateReason
        url
        bodyText
        author {
          login
        }
        comments(first: 3, orderBy: { field: UPDATED_AT, direction: DESC }) {
          nodes {
            id
            createdAt
            bodyText
            author {
              login
            }
          }
          pageInfo {
            hasPreviousPage
            hasNextPage
            startCursor
            endCursor
          }
          totalCount
        }
      }
    }
  }
`;

export const GetCommentsQuery = gql`
  query getIssueComments($issueNumber: Int!, $commentsCursor: String) {
    repository(owner: "facebook", name: "react") {
      id
      issue(number: $issueNumber) {
        id
        comments(
          first: 3
          after: $commentsCursor
          orderBy: { field: UPDATED_AT, direction: DESC }
        ) {
          nodes {
            id
            createdAt
            bodyText
            author {
              login
            }
          }
          pageInfo {
            hasPreviousPage
            hasNextPage
            startCursor
            endCursor
          }
          totalCount
        }
      }
    }
  }
`;
