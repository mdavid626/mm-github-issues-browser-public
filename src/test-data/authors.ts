import { Author } from '../types/author';

export const testAuthor1: Author = {
  login: 'testAuthor1',
};
