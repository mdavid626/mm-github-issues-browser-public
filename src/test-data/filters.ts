import { Filters } from '../types/filters';

export const testFilters1: Filters = {
  search: 'test',
  state: 'open',
};
