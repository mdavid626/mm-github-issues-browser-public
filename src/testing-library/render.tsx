/* istanbul ignore file */
import { InitialEntry, Location } from '@remix-run/router';
import { Queries, queries } from '@testing-library/dom';
import {
  RenderHookOptions,
  RenderOptions,
  render,
  renderHook,
  RenderResult,
  RenderHookResult,
} from '@testing-library/react';
import React, { ReactElement } from 'react';
import { MemoryRouter, useLocation, useNavigate } from 'react-router-dom';
import { NavigateFunction } from 'react-router/dist/lib/hooks';

type SimpleRouter = {
  location?: Location;
  navigate?: NavigateFunction;
};

type RouterResult = {
  router: SimpleRouter;
};

export const renderWithRouter = (
  component: ReactElement,
  options?: Omit<RenderOptions, 'queries'>,
  initialEntries?: InitialEntry[]
): RenderResult & RouterResult => {
  const router = {} as SimpleRouter;
  return {
    ...render(component, {
      wrapper: ({ children }) => {
        const Children = () => {
          router.location = useLocation();
          router.navigate = useNavigate();
          return children;
        };
        return (
          <MemoryRouter initialEntries={initialEntries}>
            <Children />
          </MemoryRouter>
        );
      },
      ...options,
    }),
    router,
  } as RenderResult & RouterResult;
};

export const renderHookWithRouter = <
  Result,
  Props,
  Q extends Queries = typeof queries,
  Container extends Element | DocumentFragment = HTMLElement,
  BaseElement extends Element | DocumentFragment = Container
>(
  render: (initialProps: Props) => Result,
  options?: RenderHookOptions<Props, Q, Container, BaseElement>,
  initialEntries?: InitialEntry[]
): RenderHookResult<Result, Props> & RouterResult => {
  const router = {} as SimpleRouter;
  return {
    ...renderHook(render, {
      wrapper: ({ children }) => {
        const Children = () => {
          router.location = useLocation();
          router.navigate = useNavigate();
          return children;
        };
        return (
          <MemoryRouter initialEntries={initialEntries}>
            <Children />
          </MemoryRouter>
        );
      },
      ...options,
    }),
    router,
  };
};
