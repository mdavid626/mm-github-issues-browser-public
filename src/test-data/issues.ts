import { Issue, IssueQueryResult, IssuesItem } from '../types/issue';
import { testAuthor1 } from './authors';
import { testIssueComments1 } from './issue-comments';

export const testIssue1: Issue = {
  id: 'I_kwDOAJy2Ks5d65oG',
  createdAt: '2023-02-08T08:50:14Z',
  title: 'Bug: Failed when set homepage in package.json or set PUBLIC_URL',
  state: 'CLOSED',
  number: 26126,
  url: 'https://github.com/facebook/react/issues/26126',
  body: 'test issue1 body',
  author: testAuthor1,
  comments: testIssueComments1,
};

export const testIssuesItem1: IssuesItem = {
  id: 'I_kwDOAJy2Ks5d65oG',
  createdAt: '2023-02-08T08:50:14Z',
  title: 'Bug: Failed when set homepage in package.json or set PUBLIC_URL',
  state: 'CLOSED',
  number: 26126,
  url: 'https://github.com/facebook/react/issues/26126',
  author: testAuthor1,
};

export const testIssueQueryResult1: IssueQueryResult = {
  repository: {
    issue: testIssue1,
  },
};
