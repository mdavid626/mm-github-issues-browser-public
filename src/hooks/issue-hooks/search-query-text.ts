import { Filters } from '../../types/filters';

const createSearchQueryText = (filters: Filters) => {
  const escapedSearch = filters.search.replace(/[^a-z\d\s]/gi, '');
  const search = filters.search
    ? `(in:title ${escapedSearch} OR in:body ${escapedSearch})`
    : '';
  const state = filters.state ? `state:${filters.state}` : '';
  return ['repo:facebook/react', 'type:issue', state, search]
    .filter((item) => item)
    .join(' ');
};

export default createSearchQueryText;
