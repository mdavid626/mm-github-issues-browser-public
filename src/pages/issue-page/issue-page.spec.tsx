import { cleanup, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import React from 'react';
import Footer from '../../components/footer/footer';
import Header from '../../components/header/header';
import IssueDetail from '../../components/issue-detail/issue-detail';
import { useIssue, useIssueNumber } from '../../hooks/issue-hooks/issue-hooks';
import { testIssueQueryResult1 } from '../../test-data/issues';
import { renderWithRouter } from '../../testing-library/render';
import IssuePage from './issue-page';

jest.mock('../../components/footer/footer');
jest.mock('../../components/header/header');
jest.mock('../../components/issue-detail/issue-detail');
jest.mock('../../hooks/issue-hooks/issue-hooks');
describe('issue-page', () => {
  beforeEach(() => {
    (Footer as jest.Mock).mockReturnValue(<div>footer</div>);
    (Header as jest.Mock).mockReturnValue(<div>header</div>);
    (IssueDetail as jest.Mock).mockReturnValue(<div>issue-detail</div>);
    (useIssueNumber as jest.Mock).mockReturnValue(1);
    (useIssue as jest.Mock).mockReturnValue([testIssueQueryResult1, undefined]);
  });
  afterEach(cleanup);
  afterEach(jest.resetAllMocks);

  it('should render', () => {
    const { asFragment } = renderWithRouter(<IssuePage />);
    expect(asFragment()).toMatchSnapshot();
    expect(useIssueNumber).toHaveBeenCalledWith();
    expect(useIssue).toHaveBeenCalledWith(1);
    expect(Footer).toHaveBeenCalledWith({}, {});
    expect(Header).toHaveBeenCalledWith({}, {});
    expect(IssueDetail).toHaveBeenCalledWith(
      { issue: testIssueQueryResult1.repository.issue },
      {}
    );
  });

  it('should render spinner when loading', () => {
    (useIssue as jest.Mock).mockReturnValue([undefined, undefined]);
    const { asFragment } = renderWithRouter(<IssuePage />);
    expect(asFragment()).toMatchSnapshot();
    expect(Footer).toHaveBeenCalledWith({}, {});
    expect(Header).toHaveBeenCalledWith({}, {});
    expect(IssueDetail).not.toHaveBeenCalled();
  });

  it('should show error when error', () => {
    (useIssue as jest.Mock).mockReturnValue([undefined, new Error('test error')]);
    renderWithRouter(<IssuePage />);
    expect(screen.getByText('test error')).toBeVisible();
    expect(Footer).toHaveBeenCalledWith({}, {});
    expect(Header).toHaveBeenCalledWith({}, {});
    expect(IssueDetail).not.toHaveBeenCalled();
  });

  it('should navigate back to issues when back clicked', async () => {
    const { router } = renderWithRouter(<IssuePage />, undefined, [
      '/issue/1?test=2',
    ]);
    await userEvent.click(screen.getByText('Â« back to issues'));
    expect(router.location?.pathname).toBe('/');
    expect(router.location?.search).toBe('?test=2');
  });
});
