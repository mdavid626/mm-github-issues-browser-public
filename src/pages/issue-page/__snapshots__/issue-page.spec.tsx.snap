// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`issue-page should render 1`] = `
<DocumentFragment>
  <div
    class="IssuePage"
  >
    <div>
      header
    </div>
    <div
      class="IssuePage-content"
    >
      <div
        class="IssuePage-back"
      >
        Â« back to issues
      </div>
      <div>
        issue-detail
      </div>
    </div>
    <div>
      footer
    </div>
  </div>
</DocumentFragment>
`;

exports[`issue-page should render spinner when loading 1`] = `
<DocumentFragment>
  <div
    class="IssuePage"
  >
    <div>
      header
    </div>
    <img
      alt="loading spinner"
      class="PageLoader-spinner"
      data-testid="PageLoader-spinner"
      src="spinner.svg"
    />
    <div>
      footer
    </div>
  </div>
</DocumentFragment>
`;
